packages/dds/tree/src/codec/codec.ts

```diff
@@ -0,0 +1,177 @@
+	 */
+	encode(obj: TDecoded): TEncoded;
+}
+
+/**
+ * @alpha
+ */
+export interface IDecoder<TDecoded, TEncoded> {
+	/**
+	 * Decodes `obj` from some encoded format. Typically paired with an {@link IEncoder}.
+	 */
+	decode(obj: TEncoded): TDecoded;
+}
+
+/**
+ * @alpha
+ */
+export interface IJsonCodec<TDecoded, TEncoded = JsonCompatibleReadOnly>
+	extends IEncoder<TDecoded, TEncoded>,
+		IDecoder<TDecoded, TEncoded> {}
+
+/**
+ * @remarks - TODO: We might consider using DataView or some kind of writer instead of IsoBuffer.
+ * @alpha
+ */
+export interface IBinaryCodec<TDecoded>
+	extends IEncoder<TDecoded, IsoBuffer>,
+		IDecoder<TDecoded, IsoBuffer> {}
+
+/**
+ * Contains knowledge of how to encode some in-memory type into JSON and binary formats,
+ * as well as how to decode those representations.
+ *
+ * @remarks - Codecs are typically used in shared-tree to convert data into some persisted format.
+ * For this common use case, any format for encoding that was ever actually used needs to
+ * be supported for decoding in all future code versions.
+ *
+ * Using an {@link ICodecFamily} is the recommended strategy for managing this support, keeping in
+ * mind evolution of encodings over time.
+ *
+ * @alpha
+ */
+export interface IMultiFormatCodec<TDecoded, TJsonEncoded = JsonCompatibleReadOnly> {
+	json: IJsonCodec<TDecoded, TJsonEncoded>;
+	binary: IBinaryCodec<TDecoded>;
+}
+
+/**
+ * Represents a family of codecs that can be used to encode and decode data in different formats.
+ * The family is identified by a format version, which is typically used to select the codec to use.
+ *
+ * Separating codecs into families rather than having a single codec support multiple versions (i.e. currying
+ * the `formatVersion` parameter)
+ * allows avoiding some duplicate work at encode/decode time, since the vast majority of document usage will not
+ * involve mixed format versions.
+ * @alpha
+ */
+export interface ICodecFamily<TDecoded> {
+	/**
+	 * @returns - a codec that can be used to encode and decode data in the specified format.
+	 */
+	resolve(formatVersion: number): IMultiFormatCodec<TDecoded>;
+
+	/**
+	 * @returns - an iterable of all format versions supported by this family.
+	 */
+	getSupportedFormats(): Iterable<number>;
+}
+
+/**
+ * Creates a codec family from a registry of codecs.
+ * Any codec that is not a {@link IMultiFormatCodec} will be wrapped with a default binary encoding.
+ */
+export function makeCodecFamily<TDecoded>(
+	registry: Iterable<
+		[formatVersion: number, codec: IMultiFormatCodec<TDecoded> | IJsonCodec<TDecoded>]
+	>,
+): ICodecFamily<TDecoded> {
+	const isJsonCodec = (
+		codec: IMultiFormatCodec<TDecoded> | IJsonCodec<TDecoded>,
+	): codec is IJsonCodec<TDecoded> =>
+		typeof (codec as any).encode === "function" && typeof (codec as any).decode === "function";
+	const codecs: Map<number, IMultiFormatCodec<TDecoded>> = new Map();
+	for (const [formatVersion, codec] of registry) {
```

Comment:
the body of this loop is a nice general pure function to generalize codecs. Might make sense to split it out so it can be unit tested and documented more easilly.

Reply:
i'm assuming you mean the bit about adapting if necessary (and not the map add)--done.